internal/scheduler/ranker.go

```diff
@@ -0,0 +1,192 @@
+/*
+ Copyright 2024, NVIDIA CORPORATION & AFFILIATES
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"cmp"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	maintenancev1 "github.com/Mellanox/maintenance-operator/api/v1alpha1"
+)
+
+const (
+	// InProgressRankerWeight is the weight of the InProgressRanker
+	InProgressRankerWeight = 1000
+	// LeastPendingRankerMaxWeight is the weight of the LeastPendingRanker
+	LeastPendingRankerMaxWeight = 100
+)
+
+// RankedNodeMaintenance is a wrapper for NodeMaintenance object with a rank
+type RankedNodeMaintenance struct {
+	maintenancev1.NodeMaintenance
+	Rank int
+}
+
+// CompareRanked compares between two RankedNodeMaintenance objects
+func CompareRanked(a *RankedNodeMaintenance, b *RankedNodeMaintenance) int {
```

Comment:
nit: I think it is possible to use cmp.Or() here

```
	return cmp.Or(
		cmp.Compare(a.Rank, b.Rank),
		a.CreationTimestamp.Compare(b.CreationTimestamp.Time),
		cmp.Compare(a.Spec.NodeName, b.Spec.NodeName),
		cmp.Compare(a.Spec.RequestorID, b.Spec.RequestorID),
		cmp.Compare(types.NamespacedName{Namespace: a.Namespace, Name: a.Name}.String(),
			types.NamespacedName{Namespace: b.Namespace, Name: b.Name}.String()),
	)
```

Reply:
ack, much more compact. ill add some comments additionally. lemme re-spin this bit.